File: test/fixtures/prompt_injection.py
# Smoke-test fixture: prompt_injection
#
# A Flask handler concatenates HTTP-supplied `question` directly into an OpenAI
# chat-completions prompt body with no instruction/data separation. The trust
# boundary is the request handler; the sink is the LLM prompt. A user can
# craft `question` to override the system prompt's instructions ("Ignore
# previous instructions and ..."), exfiltrate the system prompt, or pivot
# the assistant into producing arbitrary attacker-chosen output.
from flask import Flask, request
from openai import OpenAI

app = Flask(__name__)
client = OpenAI(api_key="sk-PLACEHOLDER-DO-NOT-USE")


@app.route("/ask")
def ask():
    question = request.args.get("question", "")

    # Prompt injection: user-controlled `question` is concatenated into the
    # message body with no separation from the assistant's system prompt.
    response = client.chat.completions.create(
        model="gpt-4o-mini",
        messages=[
            {"role": "system", "content": "You are a customer-service bot. Only answer questions about Acme product returns."},
            {"role": "user", "content": "User asks: " + question},
        ],
    )

    return response.choices[0].message.content

File: test/fixtures/prompt_injection.ts
// Smoke-test fixture: prompt_injection
//
// An Express handler concatenates HTTP-supplied `q` query-string parameter
// directly into an Anthropic Messages API prompt body with no
// instruction/data separation. Same shape as the Python fixture, different
// ecosystem — included so the rule pack is verified to detect agentic
// patterns in TypeScript code, not just Python.

import express, { Request, Response } from "express";
import Anthropic from "@anthropic-ai/sdk";

const app = express();
const client = new Anthropic({ apiKey: "sk-ant-PLACEHOLDER-DO-NOT-USE" });

app.get("/summarise", async (req: Request, res: Response) => {
  const query = (req.query.q ?? "") as string;

  // Prompt injection: `query` is user-controlled and is concatenated into
  // the user-role message body. An attacker who controls `q` can override
  // the system instruction or exfiltrate it.
  const message = await client.messages.create({
    model: "claude-opus-4-7",
    max_tokens: 1024,
    system: "You are a documentation summariser. Summarise the user's query.",
    messages: [
      { role: "user", content: "Please summarise: " + query },
    ],
  });

  res.send(message.content);
});

app.listen(3000);
